### PR TITLE
RTC: STM32: Add alarm support for all STM32 MCU series

### PR DESCRIPTION
**Pull request by ajarmouni-st**
This PR implements `rtc_driver_api` Alarm functions for the 2 alarms, A & B, of STM32 RTC device.
Tested on all the boards that have been added in the `tests/drivers/rtc/rtc_api/boards`.

Edit: STM32F1x SoC series is not supported.

### CODE REVIEW

File: dts/bindings/rtc/st,stm32-rtc.yaml [hunk 1]
```diff
@@ -18,3 +21,9 @@ properties:
     enum:
       - 1
       - 512
+
+  alarms-count:
+    description: |
+      Number of alarms supported by STM32 RTC device.
+      Most of STM32 MCU series have 2 RTC alarms, A & B.
+      Defaults to 0 when not declared in devicetree.
```
Comment (decsny) [inline dts/bindings/rtc/st,stm32-rtc.yaml:29]: either put a default: value here or make it required: true

**Comment by decsny** [comment]
why are there two PRs ?  #71965

File: drivers/rtc/rtc_ll_stm32.c [hunk 2]
```diff
@@ -76,46 +131,179 @@ struct rtc_stm32_config {
...
+{
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(id);
+
+	LL_RTC_EnableIT_ALR(rtc);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		LL_RTC_EnableIT_ALRA(rtc);
+	} else {
+		LL_RTC_EnableIT_ALRB(rtc);
```
Comment (gautierg-st) [inline drivers/rtc/rtc_ll_stm32.c:217]: Did you try to compile the alarm subsystem on series with only one alarm other than F1 (like C0 or F0) ? I don't think it works since all functions and variables with ALRB aren't defined.

File: drivers/rtc/rtc_ll_stm32.c [hunk 3]
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
...
+		"mask = 0x%04x", timeptr->tm_mday, timeptr->tm_wday, timeptr->tm_hour,
+		timeptr->tm_min, timeptr->tm_sec, *mask);
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int rtc_stm32_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
```
Comment (gautierg-st) [inline drivers/rtc/rtc_ll_stm32.c:712]: This function is very long, is there a way to split it in smaller pieces?

**Comment by ajarmouni-st** [comment]
> why are there two PRs ? #71965

A pure coincidence :)

File: dts/bindings/rtc/st,stm32-rtc.yaml (hunk 1 above)
Comment (ajarmouni-st) [reply dts/bindings/rtc/st,stm32-rtc.yaml:29]: There is already a default value in `rtc-device.yaml`.

**Review by gautierg-st — CHANGES_REQUESTED** [review]
I haven't reviewed everything but I have already a few comments

File: drivers/rtc/rtc_ll_stm32.c (hunk 2 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c:217]: There is a `CONFIG_SOC_SERIES_STM32F1X` just before.

Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c:217]: Good catch !!

File: drivers/rtc/rtc_ll_stm32.c [hunk 4]
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
...
+	 * that's being matched with RTC calendar to trigger alarm periodically
+	 */
+	if (id == RTC_STM32_ALRM_A) {
+		ll_mask = LL_RTC_ALMA_MASK_ALL;
+	} else {
+		ll_mask = LL_RTC_ALMB_MASK_ALL;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		if (timeptr->tm_sec > 59) {
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ll_stm32.c:785]: Have you seen 
https://github.com/zephyrproject-rtos/zephyr/blob/35429debb627943410f3c3651f267fb40f04a9cf/drivers/rtc/rtc_utils.c#L15C6-L15C33 ? unless the alarms have some additional constraint on the time fields, it can save a bit of repeated code :)

File: drivers/rtc/rtc_ll_stm32.c [hunk 5]
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
...
+		LOG_ERR("timeptr is invalid");
+		err = -EINVAL;
+		goto unlock;
+	}
+
+	if (p_rtc_alrm->user_callback != NULL) {
+		LOG_ERR("Alarm %d is busy", id);
+		err = -EBUSY;
+		goto unlock;
+	}
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ll_stm32.c:767]: It is allowed to overwrite an existing alarm, either to disable or reconfigure it :) It is not explicitly mentioned in the documentation, maybe it should be?

File: drivers/rtc/rtc_ll_stm32.c [hunk 6]
```diff
@@ -67,7 +75,54 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
...
+#else /* not CONFIG_SOC_SERIES_STM32F1X */
+#define RTC_ASYNCPRE (RTCCLK_FREQ - 1)
+#endif /* not CONFIG_SOC_SERIES_STM32F1X */
+
+/* Adjust the second sync prescaler to get 1Hz on ck_spre */
+#define RTC_SYNCPRE ((RTCCLK_FREQ / (1 + RTC_ASYNCPRE)) - 1)
+
 struct rtc_stm32_config {
+#ifdef CONFIG_RTC_ALARM
+	uint32_t alarms_count;
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ll_stm32.c:124]: A bit overkill with a u32 :) maybe uint8 or uint16 (to match with the size of id in `rtc_alarm_set_time`)

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]

File: drivers/rtc/rtc_ll_stm32.c [hunk 7]
```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
...
+#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+	LL_PWR_DisableBkUpAccess();
+#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPCR_DBP || PWR_DBPR_DBP */
+
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && defined(CONFIG_CPU_CORTEX_M4)
+	LL_C2_EXTI_ClearFlag_0_31(RTC_EXTI_LINE);
+#elif defined(CONFIG_SOC_SERIES_STM32C0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32G0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H5X)
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:1045]: #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32g0_exti)

File: drivers/rtc/rtc_ll_stm32.c [hunk 8]
```diff
@@ -67,7 +75,54 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
...
+	|| defined(CONFIG_SOC_SERIES_STM32G4X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L0X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L1X) \
+	|| defined(CONFIG_SOC_SERIES_STM32L5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H7X) \
+	|| defined(CONFIG_SOC_SERIES_STM32H5X) \
+	|| defined(CONFIG_SOC_SERIES_STM32WLX)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_17
+#endif /* CONFIG_SOC_SERIES_STM32L4X */
+#endif /* CONFIG_RTC_ALARM */
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:103]: Please use device tree to specify line number

File: drivers/rtc/rtc_ll_stm32.c [hunk 9]
```diff
@@ -67,7 +75,54 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
...
+
+#if DT_INST_CLOCKS_CELL_BY_IDX(0, 1, bus) == STM32_SRC_LSI
+/* LSI */
+#define RTCCLK_FREQ STM32_LSI_FREQ
+#else
+/* LSE */
+#define RTCCLK_FREQ STM32_LSE_FREQ
+#endif /* DT_INST_CLOCKS_CELL_BY_IDX(0, 1, bus) == STM32_SRC_LSI */
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:111]: Would be cleaner to use proper clock_control API

File: drivers/rtc/rtc_ll_stm32.c [hunk 10]
```diff
@@ -76,46 +131,179 @@ struct rtc_stm32_config {
...
+
+	return LL_RTC_IsActiveFlag_ALR(rtc);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		return LL_RTC_IsActiveFlag_ALRA(rtc);
+	}
+	if (id == RTC_STM32_ALRM_B) {
+		return LL_RTC_IsActiveFlag_ALRB(rtc);
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:203]: Can we use a specific header file to abstract these?
Similar to what is done for STM32 spi for instance.

Having those #ifdef CONFIG_SOC_SERIES_STM32F1X all over the file doesn't help in readability.

File: drivers/rtc/rtc_ll_stm32.c [hunk 11]
```diff
@@ -76,46 +131,179 @@ struct rtc_stm32_config {
...
+	ARG_UNUSED(id);
+#else
+	if (id == RTC_STM32_ALRM_A) {
+		LL_RTC_ALMA_Disable(rtc);
+	} else {
+		LL_RTC_ALMB_Disable(rtc);
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+}
+#endif /* CONFIG_RTC_ALARM */
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:264]: Good but, please put these in a dedicated header file

**Review by erwango — CHANGES_REQUESTED** [review]
Some initial comments

File: drivers/rtc/rtc_ll_stm32.c (hunk 2 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Fixed.

File: drivers/rtc/rtc_ll_stm32.c (hunk 3 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c:532]: Fixed

File: drivers/rtc/rtc_ll_stm32.c (hunk 5 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Fixed

File: drivers/rtc/rtc_ll_stm32.c (hunk 6 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Replaced by `RTC_STM32_ALARMS_COUNT` in `rtc_ll_stm32.h`

File: drivers/rtc/rtc_ll_stm32.c (hunk 7 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Fixed

File: drivers/rtc/rtc_ll_stm32.c (hunk 8 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Done.

File: drivers/rtc/rtc_ll_stm32.c (hunk 10 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Done. Dropped support for F1X series in this driver.

File: drivers/rtc/rtc_ll_stm32.c (hunk 11 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Done.

**Comment by bjarki-trackunit** [comment]
After #71784 was merged, the y2k test is now skipped automatically if the RTC correctly returns -EINVAL if the time is outside if its range, like before 2000 :) Could you try try to rebase and run the test suite without the commit excluding the y2k test?

File: dts/bindings/rtc/st,stm32-rtc.yaml (hunk 1 above)
Comment (decsny) [reply dts/bindings/rtc/st,stm32-rtc.yaml:29]: ok

**Comment by ajarmouni-st** [comment]
> After #71784 was merged, the y2k test is now skipped automatically if the RTC correctly returns -EINVAL if the time is outside if its range, like before 2000 :) Could you try try to rebase and run the test suite without the commit excluding the y2k test?

Done.

File: drivers/rtc/rtc_ll_stm32.c [hunk 12]
```diff
@@ -67,55 +78,112 @@ LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
...
 };
 
 static int rtc_stm32_configure(const struct device *dev)
 {
 	const struct rtc_stm32_config *cfg = dev->config;
-
+	LL_RTC_InitTypeDef ll_rtc_cfg;
+	int32_t hour_format     = LL_RTC_GetHourFormat(RTC);
+	int32_t sync_prescaler  = LL_RTC_GetSynchPrescaler(RTC);
+	int32_t async_prescaler = LL_RTC_GetAsynchPrescaler(RTC);
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:155]: Why changing to `int32_t` ? these API return uint32

File: drivers/rtc/rtc_ll_stm32.c
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:1]: These are too much changes for a single commit, mixing cosmetic and functional changes.
Please split changes into atomic commits.

File: drivers/rtc/rtc_ll_stm32.c [hunk 13]
```diff
@@ -22,15 +23,25 @@
...
 
 LOG_MODULE_REGISTER(rtc_stm32, CONFIG_RTC_LOG_LEVEL);
 
-#if defined(CONFIG_SOC_SERIES_STM32L1X) && !defined(RTC_SUBSECOND_SUPPORT)
-/* subsecond counting is not supported by some STM32L1x MCUs */
+#include "rtc_ll_stm32.h"
+
+/* subsecond counting is not supported by some STM32L1x MCUs (Cat.1) */
+#if (defined(CONFIG_SOC_SERIES_STM32L1X) && !defined(RTC_SUBSECOND_SUPPORT)) \
+	|| defined(CONFIG_SOC_SERIES_STM32F2X)
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:44]: Not related to alarm support, please split into a dedicated commit

File: drivers/rtc/rtc_ll_stm32.c [hunk 14]
```diff
@@ -147,28 +265,44 @@ static int rtc_stm32_init(const struct device *dev)
 
 	k_mutex_init(&data->lock);
 
-	/* Enable Backup access */
-#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+#if RTC_STM32_BACKUP_DOMAIN_WRITE_PROTECTION
 	LL_PWR_EnableBkUpAccess();
-#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPR_DBP */
+#endif /* RTC_STM32_BACKUP_DOMAIN_WRITE_PROTECTION */
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:270]: Not directly linked to alarm support, please make it a dedicated commit

File: drivers/rtc/rtc_ll_stm32.c [hunk 15]
```diff
@@ -147,28 +265,44 @@ static int rtc_stm32_init(const struct device *dev)
...
 	if (clock_control_configure(clk, (clock_control_subsys_t)&cfg->pclken[1], NULL) != 0) {
 		LOG_ERR("clock configure failed\n");
 		return -EIO;
 	}
-
+/*
+ * On STM32WBAX series, there is no bit in BCDR register to enable RTC.
+ * Enabling RTC is done directly via the RCC APB register bit.
+ */
+#ifndef CONFIG_SOC_SERIES_STM32WBAX
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:281]: Requires a dedicated commit.

File: drivers/rtc/rtc_ll_stm32.c [hunk 16]
```diff
@@ -247,22 +382,29 @@ static int rtc_stm32_set_time(const struct device *dev, const struct rtc_time *t
...
 
-	int err = k_mutex_lock(&data->lock, K_NO_WAIT);
+	if (timeptr == NULL) {
+		LOG_ERR("NULL rtc_time pointer");
+		return -EINVAL;
+	}
+
+	memset(timeptr, -1, sizeof(struct rtc_time));
+
+	err = k_mutex_lock(&data->lock, K_NO_WAIT);
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:399]: Justification ?
Likely requires a dedicated commit as well.

File: drivers/rtc/rtc_ll_stm32.c [hunk 17]
```diff
@@ -247,22 +382,29 @@ static int rtc_stm32_set_time(const struct device *dev, const struct rtc_time *t
...
 
 	if (err) {
 		return err;
 	}
 
 	if (!LL_RTC_IsActiveFlag_INITS(RTC)) {
-		/* INITS flag is set when the calendar has been initialiazed. This flag is
+		/*
+		 * INITS flag is set when the calendar has been initialiazed. This flag is
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:407]: Not related to alarm. Dedicated commit required.

File: drivers/rtc/rtc_ll_stm32.c [hunk 18]
```diff
@@ -313,15 +458,271 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
 	timeptr->tm_nsec = DIV_ROUND_CLOSEST(temp, cfg->sync_prescaler + 1);
 #else
 	timeptr->tm_nsec = 0;
-#endif
+#endif /* HW_SUBSECOND_SUPPORT */
 
-	/* unknown values */
-	timeptr->tm_yday  = -1;
-	timeptr->tm_isdst = -1;
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.c:320]: Justification ?

File: drivers/rtc/rtc_ll_stm32.h [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+		p_rtc_alarm->AlarmDateWeekDay = bin2bcd(timeptr->tm_mday);
+	}
+
+	p_rtc_alrm_time->TimeFormat = LL_RTC_TIME_FORMAT_AM_OR_24;
+
+	p_rtc_alarm->AlarmMask = ll_mask;
+
+	return 0;
+}
```
Comment (erwango) [inline drivers/rtc/rtc_ll_stm32.h:214]: This function is too long to be in a header.

Note that (unfortunately poorly named) ll_func functions used accross previous drivers are reserved to hw abstraction. 
My initial request to use these functions in previous patch set was targeting functions providing abstraction for F1 series. This is not the case anymore, so naming should be reviewed and functions can lie in driver c file when not providing abstraction.
This is not the case here, so you should use proper dedicated function names.

File: boards/st/nucleo_f722ze/nucleo_f722ze.dts [hunk 20]
```diff
@@ -174,7 +179,7 @@ zephyr_udc0: &usbotg_fs {
 
 &rtc {
 	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x10000000>,
-		<&rcc STM32_SRC_LSI RTC_SEL(2)>;
+		<&rcc STM32_SRC_LSE RTC_SEL(1)>;
```
Comment (erwango) [inline boards/st/nucleo_f722ze/nucleo_f722ze.dts:182]: Why moving to LSE ?

**Review by erwango — CHANGES_REQUESTED** [review]
One additional comment on `tests: drivers: rtc: rtc_api: support STM32 boards`:
Does it allows to cover all boards used on the bench ?

File: boards/st/nucleo_f722ze/nucleo_f722ze.dts (hunk 20 above)
Comment (ajarmouni-st) [reply boards/st/nucleo_f722ze/nucleo_f722ze.dts:182]: More precise.

Comment (erwango) [reply boards/st/nucleo_f722ze/nucleo_f722ze.dts:182]: This should be mentioned in the commit message. Also we need to keep at least one test on the LSI (could be an overlay in the test)

**Comment by ajarmouni-st** [comment]
> Does it allows to cover all boards used on the bench ?

No, a `.overlay` & a `.conf` should be added for each board for it to be tested. Also the RTC source clock must be configured in `.dts`of each board. Alternatively, a specific test case can be added for STM32 boards in `testcase.yaml` with a common overlay & `.conf`.

File: drivers/rtc/rtc_ll_stm32.c (hunk 18 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c:320]: Line 397: `memset(timeptr, -1, sizeof(struct rtc_time));`

File: drivers/rtc/rtc_ll_stm32.c (hunk 16 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c:399]: > Justification ?

We cannot operate on a NULL pointer !?

File: boards/st/nucleo_h723zg/nucleo_h723zg.dts [hunk 21]
```diff
@@ -124,7 +126,7 @@
 
 &rtc {
 	clocks = <&rcc STM32_CLOCK_BUS_APB4 0x00010000>,
-		 <&rcc STM32_SRC_LSI RTC_SEL(2)>;
+		 <&rcc STM32_SRC_LSE RTC_SEL(1)>;
```
Comment (erwango) [inline boards/st/nucleo_h723zg/nucleo_h723zg.dts:129]: Either split in a dedicated commit or add a line in commit message to explain that you change clock for xxx reason.
Not that splitting is preferred in general as it allows to easily revert in case of trouble.

**Review by erwango — CHANGES_REQUESTED** [review]
We're getting closer.

Aside from the comments, in commit "tests: drivers: rtc: rtc_api: support STM32 boards", it seems that some of the bench boards are still missing

File: drivers/rtc/rtc_ll_stm32.c (hunk 14 above)
Comment (ajarmouni-st) [reply drivers/rtc/rtc_ll_stm32.c]: Could be made clean in another PR.

Comment (erwango) [reply drivers/rtc/rtc_ll_stm32.c]: I prefer to get it now to be sure it is done ;)

**Review by erwango — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]
Nice work :)

**Comment by 721082** [comment]
Great! Is `RTC_ALARM_TIME_MASK_NSEC` support when `HW_SUBSECOND_SUPPORT==1` planned in another step?

**Comment by ajarmouni-st** [comment]
> Great! Is `RTC_ALARM_TIME_MASK_NSEC` support when `HW_SUBSECOND_SUPPORT==1` planned in another step?

Yes.

**Review by gautierg-st — APPROVED** [review]
